File: array/128_longest_consecutive_subsequence.py
# Same as C++, set is unordered so indexing can't be used
# Same approach, in brute force the only difference was that for every element in array
# We used to check if next consecutive exists using Linear Search, which used to take O(N^3) time complexity

class Solution(object):
    def longestConsecutive(self, nums):
        if (len(nums)==0):
            return 0
        longest=1
        st=set(nums)
        for i in st:
            if (i-1 not in st):
                cnt=1
                x=i
                while(x+1 in st):
                    cnt+=1
                    x+=1
                longest=max(longest,cnt)
        return longest

# Time-complexity: O(3N)
# Space-complexity: O(N)

File: array/128_longest_consecutive_subsequence.c
// GOOGLE ASKED, Optimal solution
// The trick is to only check the consecutive elements if the sequence starts from it
// Like if 102, and 101 exists then dont start checking from 102, later when 101 comes check from there
// But for that also if 100 exists then dont start from 101, start when 100 appears

class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        int n=nums.size();
        if (n==0){
            return 0;
        }
        int longest=1;
        unordered_set<int> st;
        for (int i=0;i<n;i++){
            st.insert(nums[i]);
        }
        for (auto it:st){
            if (st.find(it-1)==st.end()){
                int cnt=1;
                int x=it;
                while(st.find(x+1)!=st.end()){
                    cnt++;
                    x++;
                }
                longest=max(longest,cnt);
            }
        }
    return longest;    
    }
};

// Time-complexity: O(3N) so basically O(N)
// Space-complexity: O(N)